android-components/components/feature/search/src/main/java/mozilla/components/feature/search/telemetry/SERPTelemetryJsonParser.kt

```diff
@@ -0,0 +1,102 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.feature.search.telemetry
+
+import android.content.Context
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Deferred
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.async
+import mozilla.components.support.base.log.logger.Logger
+import mozilla.components.support.ktx.android.org.json.toList
+import mozilla.components.support.remotesettings.RemoteSettingsClient
+import mozilla.components.support.remotesettings.RemoteSettingsResult
+import mozilla.components.support.remotesettings.readOrFetch
+import org.json.JSONArray
+import org.json.JSONException
+import org.json.JSONObject
+import java.io.File
+
+/**
+ * Parse SERP Telemetry json from remote config.
+ */
+
+class SERPTelemetryJsonParser(context: Context) {
+
+    private val rootStorageDirectory: File = context.filesDir
+    private val coroutineScope = CoroutineScope(Dispatchers.IO)
+    private val searchProviders: Deferred<List<SearchProviderModel>> =
+        coroutineScope.async {
+            remoteSettingsDownloader()
+        }
+
+    suspend fun getSearchProviders(): List<SearchProviderModel> {
+        return searchProviders.await()
+    }
+
+    /**
+     * Fetches remote data from storage. If unavailable, fetches from the remote server.
+     */
+    private suspend fun remoteSettingsDownloader(): List<SearchProviderModel> {
+        val jsonFields: MutableList<SearchProviderModel> = mutableListOf()
+
+        val remoteSettingsClient = RemoteSettingsClient(
+            serverUrl = "https://firefox.settings.services.allizom.org",
+            bucketName = "main",
+            collectionName = "search-telemetry-v2",
+            storageRootDirectory = rootStorageDirectory,
+        )
+
+        val remoteRecords = remoteSettingsClient.readOrFetch()
+        if (remoteRecords is RemoteSettingsResult.Success) {
+            val records = remoteRecords.records
+            for (record in records) {
+                val result = record.fields.toSearchModelProvider()
+                if (result != null) {
+                    jsonFields.add(result)
+                }
+            }
+        }
+        return jsonFields.toList()
+    }
```

Comment:
I think this property and these too functions could be all be combined into a single function. The thing that would make more sense to me to break out (optionally) is the initialization of the `RemoteSettingsClient` into a class property, depending on whether we want to create a new object every time the function is called. even that could be done with a getter, like:

```
private val client: RemoteSettingsClient
    get() = RemoteSettingsClient(
    ....
    )
```